omegaconf/nodes.py

```diff
@@ -163,6 +163,126 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
+                object_type=type(value),
+            ),
+        )
+
+    def _wrap_node(
+        self,
+        value: Any,
+        key: Any = None,
+        is_optional: bool = True,
+        parent: Optional[Container] = None,
+        ref_type: Any = None,
+        use_type: bool = False,
+    ) -> Optional[Node]:
+        from omegaconf.omegaconf import _maybe_wrap
+
+        if value is None:
+            return None
+        else:
+            return _maybe_wrap(
+                ref_type=ref_type,
+                key=key,
+                value=value,
+                is_optional=is_optional,
+                parent=parent,  # type: ignore
+                use_type=use_type,
+            )
+
+    def validate_and_convert(self, value: Any) -> Optional[Any]:
+        from omegaconf._utils import is_structured_config
+        from omegaconf.basecontainer import BaseContainer
+
+        value_node = None
+        valid_value = False
+        value_type = type(value)
+        if isinstance(value, BaseContainer) and is_structured_config(
+            value._metadata.ref_type
+        ):
+            if value._metadata.ref_type in self.element_types:
+                return value
+            else:
+                self._raise_invalid_value(value)
+        if value_type in self.element_types:
+            self._bring_type_to_front(value_type)
+        for union_type in self.element_types:
+            try:
+                print(union_type, value)
+                value_node = self._wrap_node(
+                    value=value, ref_type=union_type, use_type=True
+                )
+                valid_value = True
+                break
+            except Exception:
+                pass
+        if not valid_value:
+            self._raise_invalid_value(value)
+        return value_node
+
+    def _raise_invalid_value(self, value: Any) -> None:
+        raise ValidationError(
+            "Value '{}' type is not in '{}'".format(value, self.element_types)
+        )
+
+    def _bring_type_to_front(self, type_: Any) -> None:
+        self.element_types.insert(
+            0, self.element_types.pop(self.element_types.index(type_))
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, ValueNode):
+            return self._value()._value() == other  # type: ignore
+        else:
+            return self._value() == other  # type: ignore
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
+
+    def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "UnionNode":
+        res = UnionNode(
+            ref_type=self._metadata.ref_type,
+            value=self._val,
+            element_types=self.element_types,
+        )
+        self._deepcopy_impl(res, memo)
+        return res
+
+    def __hash__(self) -> int:
+        return hash(self._val)
+
+    def __getstate__(self) -> Dict[str, Any]:
+        node = copy.copy(self.__dict__)
+        node["_metadata"].ref_type = Union
```

Comment:
You're right I'm changing it.

Reply:
Saw your fix, looks good, however I was wondering why you had to override `__{get,set}state__` in the first place, and it seems that all tests pass if I delete them. Could you please check whether they are really needed? (and if yes, add a test that demonstrates it)

Reply:
Ah, that is because in python 3.6 pickle doesn't like subscripted types(Dict[kt, vt], List[vt], Union[vt1, vt2]...). If you run the test with 3.6 you'll see that it fails without it.

Reply:
Oh ok thanks, resolving then! :)